libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/MapboxTripStarter.kt

```diff
@@ -0,0 +1,171 @@
+package com.mapbox.navigation.core.trip
+
+import android.annotation.SuppressLint
+import com.mapbox.android.core.permissions.PermissionsManager
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.replay.route.ReplayRouteSession
+import com.mapbox.navigation.core.trip.MapboxTripStarterExtra.MAPBOX_TRIP_STARTER_FOLLOW_DEVICE
+import com.mapbox.navigation.core.trip.MapboxTripStarterExtra.MAPBOX_TRIP_STARTER_REPLAY_ROUTE
+import com.mapbox.navigation.core.trip.session.TripSessionState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+/**
+ * This makes it simpler to start a MapboxNavigation trip session. It can be used to enable replay
+ * or to follow the device location. The [MapboxTripStarter] is not able to observe when location
+ * permissions change, so you must notify it of changes. It will check location permissions upon
+ * attach. The location permissions are not required for replay sessions.
+ *
+ * In order to share the state between an App and Android Auto, it is recommended to make this
+ * class a singleton. That will be done automatically if you use [getRegisteredInstance].
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class MapboxTripStarter private constructor() : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(MapboxTripStarterOptions.Builder().build())
+    private var replayRouteTripSession: ReplayRouteSession? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+
+    private lateinit var coroutineScope: CoroutineScope
+
+    /**
+     * Signals that the [mapboxNavigation] instance is ready for use.
+     * @param mapboxNavigation
+     */
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        this.mapboxNavigation = mapboxNavigation
+        coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
```

Comment:
I copy pasted from `UIComponent` https://github.com/mapbox/mapbox-navigation-android/blob/61415ea7ee81831b69fd698b0522bfd69119dd23/libnavui-base/src/main/java/com/mapbox/navigation/ui/base/lifecycle/UIComponent.kt#L34

I'm not seeing how it makes a difference with the tests 🤔

Reply:
`MainCoroutineRule` sets the main dispatcher so that `ThreadController` and the regular `Dispatcher.Main` behave synchronously for tests. 

Considering that kotlin classes work well, I prefer them over the `ThreadController` wrapper which seems to do nothing besides require more mockking in tests.

Reply:
On the topic of ThreadController, I have not been able to use it to resolve the issue I faced. Essentially we need to be able to replace it with  `TestCoroutineRule`

Please continue the discussion here https://mapbox.atlassian.net/browse/NAVAND-1061

Reply:
> I'm not seeing how it makes a difference with the tests 🤔

If you don't use the rule and want to pass a test scope directly. Idk why but I think @VysotskiVadim used this approach.
I personally don't have anything against creating a coroutine manually.